libfreerdp/core/update.c

```diff
@@ -854,160 +864,229 @@ static void update_write_suppress_output(wStream* s, BYTE allow, RECTANGLE_16* a
+		goto out_fail;
 
 	update_force_flush(context);
 
+	ret = TRUE;
+out_fail:
 	Stream_Release(s);
+	return ret;
 }
 
-static void update_send_surface_frame_marker(rdpContext* context, SURFACE_FRAME_MARKER* surfaceFrameMarker)
+static BOOL update_send_surface_frame_marker(rdpContext* context, SURFACE_FRAME_MARKER* surfaceFrameMarker)
 {
 	wStream* s;
 	rdpRdp* rdp = context->rdp;
+	BOOL ret = FALSE;
 
 	update_force_flush(context);
 
 	s = fastpath_update_pdu_init(rdp->fastpath);
-	update_write_surfcmd_frame_marker(s, surfaceFrameMarker->frameAction, surfaceFrameMarker->frameId);
-	fastpath_send_update_pdu(rdp->fastpath, FASTPATH_UPDATETYPE_SURFCMDS, s, FALSE);
+	if (!s)
+		return FALSE;
+
+	if (!update_write_surfcmd_frame_marker(s, surfaceFrameMarker->frameAction, surfaceFrameMarker->frameId) ||
+		!fastpath_send_update_pdu(rdp->fastpath, FASTPATH_UPDATETYPE_SURFCMDS, s, FALSE))
+		goto out_fail;
 
 	update_force_flush(context);
 
+	ret = TRUE;
+out_fail:
 	Stream_Release(s);
+	return ret;
 }
 
-static void update_send_surface_frame_bits(rdpContext* context, SURFACE_BITS_COMMAND* cmd, BOOL first, BOOL last, UINT32 frameId)
+static BOOL update_send_surface_frame_bits(rdpContext* context, SURFACE_BITS_COMMAND* cmd, BOOL first, BOOL last, UINT32 frameId)
 {
 	wStream* s;
 	rdpRdp* rdp = context->rdp;
+	BOOL ret = FALSE;
 
 	update_force_flush(context);
 
 	s = fastpath_update_pdu_init(rdp->fastpath);
-	Stream_EnsureRemainingCapacity(s, SURFCMD_SURFACE_BITS_HEADER_LENGTH + (int) cmd->bitmapDataLength + 16);
+	if (!s)
+		return FALSE;
+
+	if (!Stream_EnsureRemainingCapacity(s, SURFCMD_SURFACE_BITS_HEADER_LENGTH + (int) cmd->bitmapDataLength + 16))
+		goto out_fail;
 
 	if (first)
-		update_write_surfcmd_frame_marker(s, SURFACECMD_FRAMEACTION_BEGIN, frameId);
+	{
+		if (!update_write_surfcmd_frame_marker(s, SURFACECMD_FRAMEACTION_BEGIN, frameId))
+			goto out_fail;
+	}
 
-	update_write_surfcmd_surface_bits_header(s, cmd);
+	if (!update_write_surfcmd_surface_bits_header(s, cmd))
+		goto out_fail;
 	Stream_Write(s, cmd->bitmapData, cmd->bitmapDataLength);
 
-
 	if (last)
-		update_write_surfcmd_frame_marker(s, SURFACECMD_FRAMEACTION_END, frameId);
+	{
+		if (!update_write_surfcmd_frame_marker(s, SURFACECMD_FRAMEACTION_END, frameId))
+			goto out_fail;
+	}
 
-	fastpath_send_update_pdu(rdp->fastpath, FASTPATH_UPDATETYPE_SURFCMDS, s, cmd->skipCompression);
+	ret = fastpath_send_update_pdu(rdp->fastpath, FASTPATH_UPDATETYPE_SURFCMDS, s, cmd->skipCompression);
 
 	update_force_flush(context);
 
+out_fail:
 	Stream_Release(s);
+	return ret;
 }
 
-static void update_send_frame_acknowledge(rdpContext* context, UINT32 frameId)
+static BOOL update_send_frame_acknowledge(rdpContext* context, UINT32 frameId)
 {
 	wStream* s;
 	rdpRdp* rdp = context->rdp;
 
 	if (rdp->settings->ReceivedCapabilities[CAPSET_TYPE_FRAME_ACKNOWLEDGE])
 	{
+		BOOL ret;
 		s = rdp_data_pdu_init(rdp);
+		if (!s)
```

Comment:
rdp_data_pdu_init initializes the stream with 2048 byte. And we only write 4 byte plus headers. I think the check isn't necessary here.